app/src/main/java/com/github/warnastrophy/core/data/repository/HybridContactRepository.kt

```diff
@@ -1,76 +1,146 @@
+        remoteOperation = { remote.editContact(userId, contactID, newContact) })
   }
 
   override suspend fun deleteContact(userId: String, contactID: String): Result<Unit> {
-    val localDelete = local.deleteContact(userId, contactID)
-    if (localDelete.isSuccess) remote.deleteContact(userId, contactID)
+    return updateBothRepositories(
+        localOperation = { local.deleteContact(userId, contactID) },
+        remoteOperation = { remote.deleteContact(userId, contactID) })
+  }
 
-    return localDelete
+  /**
+   * Synchronizes the contacts from the remote repository to the local repository. This method will
+   * add any contacts in the remote repository that do not exist in the local repository.
+   *
+   * @param userId The user ID whose contacts need to be synchronized.
+   * @param remoteContacts The list of contacts fetched from the remote repository.
+   */
+  private suspend fun syncRemoteToLocal(userId: String, remoteContacts: List<Contact>) {
+    try {
+      val localContacts = local.getAllContacts(userId).getOrNull().orEmpty()
+      val localIds = localContacts.map { it.id }.toSet()
+
+      for (contact in remoteContacts) {
+        if (contact.id !in localIds) {
+          local.addContact(userId, contact)
+        }
+      }
+    } catch (e: Exception) {
+      Log.w(hybridRepositoryTag, "Sync to local failed: ${e.localizedMessage}")
+    }
   }
 
-  private suspend fun localContactsMerge(userId: String, remoteContacts: List<Contact>) {
-    val localContacts = local.getAllContacts(userId).getOrNull().orEmpty()
-    val localIds = localContacts.map { it.id }.toSet()
+  /**
+   * A utility method that performs an operation on both the local and remote repositories. First,
+   * the local repository is updated. If successful, it checks if the remote repository is available
+   * and attempts to update it. If the remote operation fails, the local result is returned.
+   *
+   * @param localOperation A suspending function that performs the local operation.
+   * @param remoteOperation A suspending function that performs the remote operation.
+   * @return A [Result] indicating success or failure of the operation.
+   */
+  private suspend fun updateBothRepositories(
+      localOperation: suspend () -> Result<Unit>,
+      remoteOperation: suspend () -> Result<Unit>
+  ): Result<Unit> {
 
-    for (c in remoteContacts) {
-      if (c.id !in localIds) {
-        local.addContact(userId, c)
-      }
+    val localResult = localOperation()
+
+    if (localResult.isFailure) {
+      return localResult
     }
+
+    val shouldUpdateRemote = syncMutex.withLock { isRemoteAvailable }
+    if (!shouldUpdateRemote) return localResult
+
+    try {
+      remoteOperation()
+      syncMutex.withLock { isRemoteAvailable = true }
+    } catch (e: Exception) {
+      syncMutex.withLock { isRemoteAvailable = false }
+      Log.w(hybridRepositoryTag, "Remote update failed: ${e.localizedMessage}")
+    }
```

Comment:
You correctly handled the `Failure` case for `localOperation` but not for `remoteOperation`. Indeed, while you handled the `Exception` case with `try-catch` block, if `remoteOperation` fails (i.e return `Result<Failure>`), it will be treated as correct, because not an exception. Here a suggestion:

```suggestion
        val remoteResult = remoteOperation()
    if (remoteResult.isFailure) {
        syncMutex.withLock { isRemoteAvailable = false }
        val exception = remoteResult.exceptionOrNull()
        Log.w(hybridRepositoryTag, "Remote update failed, marking remote as unavailable: ${exception?.localizedMessage}")

        // Return the remote failure to the caller
        return remoteResult
    }
```